{"cells":[{"cell_type":"markdown","metadata":{},"source":["# Bio Computing Course"]},{"cell_type":"markdown","metadata":{},"source":["# Instructions\n","### Welcome to the Python and Biological Concepts practice notebook! This set of exercises will help you apply your understanding of basic Python programming while exploring biological concepts. Please read each question carefully and use the provided code cells to write your solutions.\n","\n","# General Guidelines:\n","### Read Each Question Carefully: Make sure you understand what is being asked before you start coding. Pay close attention to the input and output requirements.\n","\n","### Write Clear and Efficient Code: Aim to write code that is both correct and efficient. Use basic Python concepts such as loops, conditionals, lists, dictionaries, and strings as needed."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["dna_sequence = \"ATGCGTACGTT\""]},{"cell_type":"markdown","metadata":{},"source":["### 1. String Manipulation - DNA Sequence\n","#### Question: Given a DNA sequence \"ATGCGTACGTT\", write a Python function to count the number of times the nucleotide 'G' appears."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 2. List - DNA Codons\n","#### Question: Create a list of all codons (3-letter combinations) in the DNA sequence \"ATGCGTACGTT\"."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 3. Dictionary - Codon to Amino Acid Mapping\n","#### Question: Create a dictionary that maps codons to their corresponding amino acids for the following: 'ATG': 'Methionine', 'CGT': 'Arginine', 'TAC': 'Tyrosine', 'GTT': 'Valine'."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["codon_to_amino_acid = {\n","    'ATG': 'Methionine',\n","    'CGT': 'Arginine',\n","    'TAC': 'Tyrosine',\n","    'GTT': 'Valine'\n","}"]},{"cell_type":"markdown","metadata":{},"source":["### 4. Conditional Statements - DNA Base Identification\n","#### Question: Write a function that identifies if a base is a purine (A, G) or pyrimidine (C, T)."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 5. Loops - Counting Bases\n","#### Question: Write a function to count the number of each base in the DNA sequence \"ATGCGTACGTT\"."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 6. String Slicing - mRNA from DNA\n","\n","#### Question: Convert the DNA sequence \"ATGCGTACGTT\" to its mRNA sequence.\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 7. List - Reverse Complement of DNA\n","#### Question: Write a function that returns the reverse complement of a DNA sequence \"ATGCGTACGTT\"."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 8. Dictionary - Transcription Mapping\n","\n","#### Question: Create a dictionary to map each DNA base to its RNA complement (A->U, T->A, C->G, G->C) and use it to transcribe \"ATGCGTACGTT\".\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 9. Conditional Statements - DNA Sequence Validity\n","\n","#### Question: Write a function to check if a given sequence is a valid DNA sequence (only contains A, T, C, G).\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 10. Loops - Translation of Codons\n","\n","#### Question: Write a function to translate the DNA sequence \"ATGCGTACGTT\" into a protein sequence using the codon dictionary provided."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 11. String - Counting GC Content\n","\n","#### Question: Write a function to calculate the GC content of a DNA sequence \"ATGCGTACGTT\"."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 12. List - Extract Exons\n","\n","#### Question: Given a list of exon positions [(0, 3), (4, 7)], extract the exons from the DNA sequence \"ATGCGTACGTT\"."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 13. Dictionary - Frequency of Codons\n","\n","#### Question: Write a function to count the frequency of each codon in the DNA sequence \"ATGCGTACGTT\"."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 14. String Slicing - Subsequence Check\n","#### Question: Write a function to check if the sequence \"CGTAC\" is a subsequence of \"ATGCGTACGTT\"."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 15. Loops - Finding Start Codon\n","#### Question: Write a function to find the position of the first start codon (ATG) in the DNA sequence \"ATGCGTACGTT\"."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 16. Dictionary - Counting Nucleotides\n","#### Question: Write a function to count the occurrence of each nucleotide in the DNA sequence \"ATGCGTACGTT\" using a dictionary."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 17. List - Split DNA Sequence into Codons\n","#### Question: Write a function to split the DNA sequence \"ATGCGTACGTT\" into codons and store them in a list."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 18. Conditional Statements - Check Palindrome\n","#### Question: Write a function to check if the DNA sequence \"ATGCGCAT\" is a palindrome."]},{"cell_type":"code","execution_count":169,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["DNA : ATTA\n","DNA Seq is Palinddrome\n"]}],"source":["def check_palindrome(x : str) -> bool:\n","    #x = \"ATTA\"\n","\n","    is_palindrome = True\n","    n = len(x)\n","    for i in range(n//2):\n","        if x[i] != x[n-i-1]:\n","            is_palindrome = False\n","        \n","    return is_palindrome\n","\n","dna_sequence = \"ATTA\"\n","\n","print(f\"DNA : {dna_sequence}\")\n","\n","if check_palindrome(dna_sequence) == True:\n","    print(\"DNA Seq is Palinddrome\")\n","else :\n","    print(\"DNA seq is NOT palindrome\")\n","\n","\n","\n"]},{"cell_type":"markdown","metadata":{},"source":["### 19. Loops - Complementary DNA Strand\n","#### Question: Write a function to generate the complementary DNA strand for \"ATGCGTACGTT\"."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 20. Dictionary - Counting Codons in mRNA\n","#### Question: Write a function to count the frequency of each codon in the mRNA sequence \"AUGCGUACGUU\" using a dictionary."]},{"cell_type":"code","execution_count":29,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["{'AAA': 0, 'AAU': 0, 'AAG': 0, 'AAC': 0, 'AUA': 0, 'AUU': 0, 'AUG': 0, 'AUC': 0, 'AGA': 0, 'AGU': 0, 'AGG': 0, 'AGC': 0, 'ACA': 0, 'ACU': 0, 'ACG': 0, 'ACC': 0, 'UAA': 0, 'UAU': 0, 'UAG': 0, 'UAC': 0, 'UUA': 0, 'UUU': 0, 'UUG': 0, 'UUC': 0, 'UGA': 0, 'UGU': 0, 'UGG': 0, 'UGC': 0, 'UCA': 0, 'UCU': 0, 'UCG': 0, 'UCC': 0, 'GAA': 0, 'GAU': 0, 'GAG': 0, 'GAC': 0, 'GUA': 0, 'GUU': 0, 'GUG': 0, 'GUC': 0, 'GGA': 0, 'GGU': 0, 'GGG': 0, 'GGC': 0, 'GCA': 0, 'GCU': 0, 'GCG': 0, 'GCC': 0, 'CAA': 0, 'CAU': 0, 'CAG': 0, 'CAC': 0, 'CUA': 0, 'CUU': 0, 'CUG': 0, 'CUC': 0, 'CGA': 0, 'CGU': 0, 'CGG': 0, 'CGC': 0, 'CCA': 0, 'CCU': 0, 'CCG': 0, 'CCC': 0}\n","{'AAA': 0,\n"," 'AAC': 0,\n"," 'AAG': 0,\n"," 'AAU': 0,\n"," 'ACA': 0,\n"," 'ACC': 0,\n"," 'ACG': 1,\n"," 'ACU': 0,\n"," 'AGA': 0,\n"," 'AGC': 0,\n"," 'AGG': 0,\n"," 'AGU': 0,\n"," 'AUA': 0,\n"," 'AUC': 0,\n"," 'AUG': 1,\n"," 'AUU': 0,\n"," 'CAA': 0,\n"," 'CAC': 0,\n"," 'CAG': 0,\n"," 'CAU': 0,\n"," 'CCA': 0,\n"," 'CCC': 0,\n"," 'CCG': 0,\n"," 'CCU': 0,\n"," 'CGA': 0,\n"," 'CGC': 0,\n"," 'CGG': 0,\n"," 'CGU': 1,\n"," 'CUA': 0,\n"," 'CUC': 0,\n"," 'CUG': 0,\n"," 'CUU': 0,\n"," 'GAA': 0,\n"," 'GAC': 0,\n"," 'GAG': 0,\n"," 'GAU': 0,\n"," 'GCA': 0,\n"," 'GCC': 0,\n"," 'GCG': 0,\n"," 'GCU': 0,\n"," 'GGA': 0,\n"," 'GGC': 0,\n"," 'GGG': 0,\n"," 'GGU': 0,\n"," 'GUA': 0,\n"," 'GUC': 0,\n"," 'GUG': 0,\n"," 'GUU': 0,\n"," 'UAA': 0,\n"," 'UAC': 0,\n"," 'UAG': 0,\n"," 'UAU': 0,\n"," 'UCA': 0,\n"," 'UCC': 0,\n"," 'UCG': 0,\n"," 'UCU': 0,\n"," 'UGA': 0,\n"," 'UGC': 0,\n"," 'UGG': 0,\n"," 'UGU': 0,\n"," 'UUA': 0,\n"," 'UUC': 0,\n"," 'UUG': 0,\n"," 'UUU': 0}\n"]}],"source":["from pprint import pprint\n","\n","nuc = \"AUGC\"\n","\n","codon_list = [(char_1+char_2+char_3,0) for char_1 in nuc for char_2 in nuc for char_3 in nuc]\n","#print(codon_list)\n","\n","count = dict(codon_list)\n","print(count)\n","\n","dna_seq = \"AUGCGUACGUU\"\n","n = len(dna_seq)\n","for i in range(0,n,3):\n","\n","    # to handle case when seq length is not a multiple of 3\n","    if len(dna_seq[i:i+3]) < 3:\n","        break\n","\n","    codon = dna_seq[i:i+3]\n","    count[codon] += 1\n","\n","pprint(count)\n","\n","\n"]},{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":["codon_to_amino_acid = {\n","    'ATG': 'Methionine', 'CGT': 'Arginine', 'TAC': 'Tyrosine', 'GTT': 'Valine',\n","    'TAA': 'Stop', 'TAG': 'Stop', 'TGA': 'Stop'\n","}"]},{"cell_type":"markdown","metadata":{},"source":["### 21. Dictionary & List - DNA to Protein Translation with a Stop Codon\n","\n","#### Question: Given a dictionary mapping codons to amino acids and a DNA sequence \"ATGCGTTAA\", write a function to translate the DNA sequence into a protein sequence, stopping translation at the first stop codon ('TAA', 'TAG', 'TGA')."]},{"cell_type":"code","execution_count":36,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["DNA : AUGCGUACGUU\n","Protein : Methionine-Arginine-Threonine\n"]}],"source":["codon_to_amino_acid = {\n","        'AUG': 'Methionine', 'UUU': 'Phenylalanine', 'UUC': 'Phenylalanine', 'UUA': 'Leucine', 'UUG': 'Leucine',\n","        'CUU': 'Leucine', 'CUC': 'Leucine', 'CUA': 'Leucine', 'CUG': 'Leucine', 'AUU': 'Isoleucine', 'AUC': 'Isoleucine',\n","        'AUA': 'Isoleucine', 'GUU': 'Valine', 'GUC': 'Valine', 'GUA': 'Valine', 'GUG': 'Valine', 'UCU': 'Serine',\n","        'UCC': 'Serine', 'UCA': 'Serine', 'UCG': 'Serine', 'CCU': 'Proline', 'CCC': 'Proline', 'CCA': 'Proline',\n","        'CCG': 'Proline', 'ACU': 'Threonine', 'ACC': 'Threonine', 'ACA': 'Threonine', 'ACG': 'Threonine', 'GCU': 'Alanine',\n","        'GCC': 'Alanine', 'GCA': 'Alanine', 'GCG': 'Alanine', 'UAU': 'Tyrosine', 'UAC': 'Tyrosine', 'UAA': 'Stop',\n","        'UAG': 'Stop', 'CAU': 'Histidine', 'CAC': 'Histidine', 'CAA': 'Glutamine', 'CAG': 'Glutamine', 'AAU': 'Asparagine',\n","        'AAC': 'Asparagine', 'AAA': 'Lysine', 'AAG': 'Lysine', 'GAU': 'Aspartic Acid', 'GAC': 'Aspartic Acid',\n","        'GAA': 'Glutamic Acid', 'GAG': 'Glutamic Acid', 'UGU': 'Cysteine', 'UGC': 'Cysteine', 'UGA': 'Stop',\n","        'UGG': 'Tryptophan', 'CGU': 'Arginine', 'CGC': 'Arginine', 'CGA': 'Arginine', 'CGG': 'Arginine', 'AGU': 'Serine',\n","        'AGC': 'Serine', 'AGA': 'Arginine', 'AGG': 'Arginine', 'GGU': 'Glycine', 'GGC': 'Glycine', 'GGA': 'Glycine',\n","        'GGG': 'Glycine'\n","    }\n","stop_codon = ['TAA','TAG','TGA']\n","\n","dna_seq = \"AUGCGUACGUU\"\n","n = len(dna_seq)\n","\n","protein = \"\"\n","\n","\n","for i in range(0,n,3):\n","    \n","    codon = dna_seq[i:i+3]\n","\n","    # if codon len is < 3 then its invalid\n","    if len(codon) < 3 :        \n","        break\n","\n","    # if codon is stop codon , add stop in protein and break\n","    if codon in stop_codon:\n","        protein += \"-stop\"\n","        break\n","\n","    # continue adding protein corresponding to codon\n","\n","    amino_acid = codon_to_amino_acid[codon]\n","\n","    protein +=  amino_acid + \"-\"\n","\n","print(f\"DNA : {dna_seq}\")\n","#print(f\"Protein : {protein}\")\n","print(f\"Protein : {protein[:-1]}\")\n","\n","\n","\n"]},{"cell_type":"markdown","metadata":{},"source":["### 22. Set Operations - Unique Nucleotide Combinations\n","\n","#### Question: Write a function that finds all unique nucleotide combinations of length 2 in a DNA sequence \"ATGCGTACGTT\"."]},{"cell_type":"code","execution_count":42,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["['AT', 'GC', 'GT', 'AC']\n"]}],"source":["from pprint import pprint\n","\n","unique_nu_2 = []\n","dna_seq = \"ATGCGTACGTT\"\n","n = len(dna_seq)\n","\n","for i in range(0,n,2):\n","\n","    nuc_pair = dna_seq[i:i+2]\n","\n","    # To handel case when length of seq is odd & last nuc will not form pair\n","    if len(nuc_pair) < 2:\n","        break\n","\n","    # just add to unique nuc pair list\n","    if nuc_pair not in unique_nu_2:\n","        unique_nu_2.append(nuc_pair)\n","\n","\n","pprint(unique_nu_2)\n","\n","\n"]},{"cell_type":"code","execution_count":47,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["{'GT', 'GC', 'AT', 'AC'}\n"]}],"source":["from pprint import pprint\n","\n","unique_nu_2 = set()\n","dna_seq = \"ATGCGTACGTT\"\n","n = len(dna_seq)\n","\n","for i in range(0,n,2):\n","\n","    nuc_pair = dna_seq[i:i+2]\n","\n","    # To handel case when length of seq is odd & last nuc will not form pair\n","    if len(nuc_pair) < 2:\n","        break\n","\n","    # just add to unique nuc pair list\n","    \n","    unique_nu_2.add(nuc_pair)\n","\n","\n","pprint(unique_nu_2)\n","\n","\n"]},{"cell_type":"markdown","metadata":{},"source":["### 23. Nested Loops - Finding All ORFs\n","#### Question: Write a function to find all open reading frames (ORFs) in the DNA sequence \"ATGCGTACGTTATGCGTTAA\" starting with 'ATG' and ending with a stop codon."]},{"cell_type":"code","execution_count":99,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["ORF 1 : CGTACTGG\n","ORF 2 : CGTACG\n","ORF 3 : CGTACC\n","ORF 4 : CGTAC\n"]}],"source":["# using regular expressions \n","\n","import re \n","\n","seq = \"ATGCGTACTGGTAAATGCGTACGTAGATGCGTACCTGAATGCGTACTGATGA\" \n","\n","stop_codon = ['TAA','TAG','TGA']\n","pattern = r\"ATG((?:[ATGC]{1})*?)(TAA|TAG|TGA)\"\n","\n","pattern = re.compile(pattern , re.IGNORECASE)\n","matches = pattern.finditer(seq)\n","\n","for i,match in enumerate(matches,1):    \n","    #print(match.group(0))\n","    print(f\"ORF {i} : {match.group(1)}\")\n","    \n","    "]},{"cell_type":"markdown","metadata":{},"source":["### 24. Regular Expressions - Validating DNA Sequence\n","#### Question: Use a regular expression to validate if a given DNA sequence \"ATGCGTACGTT\" only contains valid nucleotides (A, T, C, G)."]},{"cell_type":"code","execution_count":105,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["DNA Sequence : ATGCATGfCAGgT\n","DNA is INVALID\n"]}],"source":["import re\n","seq = \"ATGCATGfCAGgT\"\n","\n","\n","pattern = r\"[^ATGC]\"\n","\n","pattern = re.compile(pattern , re.IGNORECASE)\n","matches = pattern.findall(seq)\n","\n","print(f\"DNA Sequence : {seq}\")\n","\n","if matches:\n","    print(\"DNA is INVALID\")\n","\n","else:\n","    print(\"DNA is VALID\")\n","\n"]},{"cell_type":"markdown","metadata":{},"source":["### 25. Recursion - Calculate GC Content Recursively\n","#### Question: Write a recursive function to calculate the GC content of a DNA sequence \"ATGCGTACGTT\"."]},{"cell_type":"code","execution_count":125,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["DNA : ATGCGTACGTT\n","GC Count : 5\n"]}],"source":["def count_gc(seq,i):\n","    if i == 0 and (seq[i] in ['G','C']):        \n","        return 1\n","    \n","\n","    if i == 0 and (seq[i] not in ['G','C']):\n","        return 0\n","    \n","    if seq[i] in ['G','C']:        \n","        return 1 + count_gc(seq,i-1)\n","    else: \n","        return count_gc(seq,i-1)\n","\n","seq = \"ATGCGTACGTT\"\n","n = len(seq)\n","\n","print(f\"DNA : {seq}\")\n","print(f\"GC Count : {count_gc(seq,n-1)}\")"]},{"cell_type":"markdown","metadata":{},"source":["### 27. List & Set - Finding Unique and Common Codons\n","#### Question: Write a function to find unique and common codons between two DNA sequences \"ATGCGTACGTT\" and \"ATGCGTGTGTA\"."]},{"cell_type":"code","execution_count":139,"metadata":{},"outputs":[{"data":{"text/plain":["{'ACG', 'ATG', 'CGT', 'GTG'}"]},"execution_count":139,"metadata":{},"output_type":"execute_result"}],"source":["def find_codon(dna):\n","    codon_list = set()\n","    n = len(dna)\n","\n","    for i in range(0,n,3):\n","        codon = dna[i:i+3]\n","        if len(codon) < 3:\n","            break\n","        \n","        codon_list.add(codon)\n","\n","    return codon_list\n","\n","def find_unique_codon(dna1 ,dna2):\n","\n","    codon_set_1 = find_codon(dna1)\n","    codon_set_2 = find_codon(dna2)\n","\n","    unique_common_codon =  codon_set_1.union(codon_set_2)\n","\n","    return unique_common_codon\n","\n","dna1 = \"ATG CGT ACG TT\"\n","dna2 = \"ATG CGT GTG TA\"\n","\n","find_unique_codon(dna1, dna2)"]},{"cell_type":"markdown","metadata":{},"source":["### 28. List Comprehension - Transcribe Multiple DNA Sequences\n","#### Question: Write a function to transcribe a list of DNA sequences [\"ATGCGT\", \"GATTACA\", \"CGTACG\"] into their respective mRNA sequences using list comprehension."]},{"cell_type":"code","execution_count":141,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["['ATGCGT', 'GATTACA', 'CGTACG']\n","['AUGCGU', 'GAUUACA', 'CGUACG']\n"]}],"source":["dna_seq = [\"ATGCGT\", \"GATTACA\", \"CGTACG\"]\n","\n","rna_seq = [dna.replace('T','U') for dna in dna_seq]\n","\n","print(dna_seq)\n","print(rna_seq)"]},{"cell_type":"markdown","metadata":{},"source":["### 29. Nested Loops - Counting Overlapping Codons\n","#### Question: Write a function to count the number of times a specific codon \"CGT\" appears in a DNA sequence \"ATGCGTACGTT\" including overlapping occurrences."]},{"cell_type":"code","execution_count":146,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["DNA : ATGCGTACGTT\n","codon : CGT\n","overlapping : 2\n"]}],"source":["codon = \"CGT\"\n","dna = \"ATGCGTACGTT\"\n","n = len(dna)\n","\n","count = 0\n","\n","for i in range(n):\n","    local_codon = dna[i:i+3]\n","\n","    if len(local_codon) < 3:\n","        break\n","\n","    if local_codon == codon:\n","        count += 1\n","\n","print(f\"DNA : {dna}\")\n","print(f\"codon : {codon}\")\n","print(f\"overlapping : {count}\")\n","\n","\n","    "]},{"cell_type":"markdown","metadata":{},"source":["### 30. List & Dictionary - Grouping DNA Sequences by GC Content\n","#### Question: Write a function to group DNA sequences [\"ATGCGT\", \"GATTACA\", \"CGTACG\", \"TATATA\"] into categories based on GC content: \"High GC\" (>50%), \"Moderate GC\" (30-50%), \"Low GC\" (<30%)."]},{"cell_type":"code","execution_count":162,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["High GC DNA: \n","ATGCGT CGTACG \n","\n","Moderate GC DNA: \n","\n","\n","low GC DNA: \n","GATTACA TATATA \n","\n"]}],"source":["dna_list = [\"ATGCGT\", \"GATTACA\", \"CGTACG\", \"TATATA\"]\n","\n","high_gc_dna_list = [dna for dna in dna_list if (count_gc(dna,len(dna) -1 )/len(dna) >= .5)]\n","\n","moderate_gc_dna_list = [dna for dna in dna_list if (count_gc(dna,len(dna) -1 )/len(dna) >= .3) and (count_gc(dna,len(dna) -1 )/len(dna) < .5 )]\n","\n","low_gc_dna_list = [dna for dna in dna_list if (count_gc(dna,len(dna) -1 )/len(dna) < .3)]\n","\n","print(\"High GC DNA: \")\n","[print(dna,end=\" \") for dna in high_gc_dna_list]\n","print(\"\\n\")\n","\n","print(\"Moderate GC DNA: \")\n","[print(dna,end=\" \") for dna in moderate_gc_dna_list]\n","print(\"\\n\")\n","\n","print(\"low GC DNA: \")\n","[print(dna,end=\" \") for dna in low_gc_dna_list]\n","print(\"\\n\")\n","\n"]},{"cell_type":"markdown","metadata":{},"source":["### 31. List & Conditional Statements - Counting Transitions and Transversions\n","#### Question: Write a function that counts the number of transitions (purine to purine, pyrimidine to pyrimidine) and transversions (purine to pyrimidine, vice versa) between two DNA sequences \"ATGCGTAC\" and \"ATGCGTAG\"."]},{"cell_type":"code","execution_count":168,"metadata":{},"outputs":[{"data":{"text/plain":["(7, 1)"]},"execution_count":168,"metadata":{},"output_type":"execute_result"}],"source":["\n","def count_transitions_transversions(dna1, dna2):\n","    purines = ['A','G']\n","    pyrimidines = ['T','C']\n","\n","    transition = 0\n","    transversions = 0\n","\n","    for nuc_1, nuc_2 in zip(dna1,dna2):\n","        if (nuc_1 in purines and nuc_2 in purines ) or (nuc_1 in pyrimidines and nuc_2 in pyrimidines ):\n","            transition += 1\n","\n","        if (nuc_1 in purines and nuc_2 in pyrimidines ) or (nuc_1 in pyrimidines and nuc_2 in purines ):\n","            transversions += 1\n","\n","    return (transition,transversions)\n","\n","dna_1 = \"ATGCGTAC\" \n","dna_2 = \"ATGCGTAG\"\n","count_transitions_transversions(dna_1, dna_2)"]},{"cell_type":"code","execution_count":166,"metadata":{},"outputs":[{"data":{"text/plain":["False"]},"execution_count":166,"metadata":{},"output_type":"execute_result"}],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 32. List & Dictionary - Identifying Palindromic Sequences\n","#### Question: Write a function to find all palindromic sequences of length 4 in a DNA sequence \"ATGCGTACGTACGCGT\"."]},{"cell_type":"code","execution_count":172,"metadata":{},"outputs":[{"data":{"text/plain":["['ATTA']"]},"execution_count":172,"metadata":{},"output_type":"execute_result"}],"source":["def find_4_len_palin(dna):\n","\n","\n","    n = len(dna)\n","    palin_list = []\n","\n","    for i in range(n):\n","        sm_dna = dna[i:i+4]\n","        \n","        if(len(sm_dna) < 4):            \n","            break\n","\n","        if check_palindrome(sm_dna) == True:\n","            palin_list.append(sm_dna)\n","\n","    return palin_list\n","\n","\n","\n","dna = \"ATTAGTACGTACGCGT\"\n","\n","find_4_len_palin(dna)\n","        \n","\n"]},{"cell_type":"markdown","metadata":{},"source":["### 33. String & List - DNA Mutation Simulation\n","#### Question: Write a function to simulate a point mutation in a DNA sequence \"ATGCGTAC\" by randomly replacing one nucleotide with another."]},{"cell_type":"code","execution_count":11,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["DNA : ATGCGTAC\n","Random mutation at 5\n","Mutated DNA : ATGCGAAC\n"]}],"source":["import random\n","\n","def random_nuc(nuc): \n","    nuc_list = ['A','T','G','C']\n","    nuc_list.remove(nuc)\n","    return nuc_list[random.randint(0,2)]\n","\n","\n","\n","\n","dna = \"ATGCGTAC\"\n","n = len(dna)\n","\n","random_index = random.randint(0,n-1)\n","\n","print(f\"DNA : {dna}\")\n","print(f\"Random mutation at {random_index}\")\n","\n","mutated_dna = dna[0:random_index] + random_nuc(dna[random_index]) + dna[random_index+1 : ]\n","\n","print(f\"Mutated DNA : {mutated_dna}\")\n"," \n"]},{"cell_type":"markdown","metadata":{},"source":["### 34. Loops & Dictionary - Counting Nucleotides in Multiple Sequences\n","#### Question: Write a function to count the occurrence of each nucleotide in a list of DNA sequences [\"ATG\", \"CGT\", \"TAC\", \"GTT\"]."]},{"cell_type":"code","execution_count":19,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["DNA : ATG\n","Count A : 1\n","Count T : 1\n","Count G : 1\n","Count C : 0\n","\n","DNA : CGT\n","Count A : 0\n","Count T : 1\n","Count G : 1\n","Count C : 1\n","\n","DNA : TAC\n","Count A : 1\n","Count T : 1\n","Count G : 0\n","Count C : 1\n","\n","DNA : GTT\n","Count A : 0\n","Count T : 2\n","Count G : 1\n","Count C : 0\n","\n"]}],"source":["nucleotides = ['A','T','G','C']\n","\n","def count_nuc(dna):\n","    count = {'A':0,'T':0,'G':0,'C':0}\n","\n","    for key in count.keys():\n","        count[key] = dna.count(key)\n","\n","    return count\n","\n","\n","def print_count_nuc(count):\n","\n","    nucleotides = ['A','T','G','C']\n","\n","    \n","    for nuc in nucleotides:\n","        print(f\"Count {nuc} : {count[nuc]}\")\n","\n","\n","dna_list = [\"ATG\", \"CGT\", \"TAC\", \"GTT\"]\n","\n","for dna in dna_list:\n","    print(f\"DNA : {dna}\")\n","    print_count_nuc(count_nuc(dna))\n","    print()\n","\n","    \n","\n"]},{"cell_type":"markdown","metadata":{},"source":["### 35. String Manipulation - GC Skew Calculation\n","#### Question: Write a function to calculate the GC skew (G - C / G + C) at each position in the DNA sequence \"ATGCGTACGTT\"."]},{"cell_type":"code","execution_count":27,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["DNA : ATGCGTACGTT\n","GC skew : 4.333\n"]}],"source":["nucleotides = ['A','T','G','C']\n","\n","def count_nuc(dna):\n","    count = {'A':0,'T':0,'G':0,'C':0}\n","\n","    for key in count.keys():\n","        count[key] = dna.count(key)\n","\n","    return count\n","\n","dna = \"ATGCGTACGTT\"\n","count_all = count_nuc(dna)\n","\n","gc_skew = count_all['G'] - count_all['C'] / count_all['G'] + count_all['C']\n","\n","print(f\"DNA : {dna}\")\n","print(f\"GC skew : {gc_skew:.3f}\")\n"]},{"cell_type":"markdown","metadata":{},"source":["### 36. List & Set - Finding Common and Unique Nucleotides\n","#### Question: Write a function to find common and unique nucleotides between two DNA sequences \"ATGCGT\" and \"GATTACA\"."]},{"cell_type":"code","execution_count":32,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["dna 1 : ATT\n","dna 2 : ATGC\n","common & Unique Nucleotides are :  A T "]}],"source":["dna_1 = \"ATT\" \n","dna_2 = \"ATGC\"\n","\n","common_unique = set(dna_1).intersection(set(dna_2))\n","\n","print(f\"dna 1 : {dna_1}\")\n","print(f\"dna 2 : {dna_2}\")\n","\n","print(f\"common & Unique Nucleotides are : \",end=\" \")\n","for nuc in common_unique:\n","    print(f\"{nuc}\",end=\" \")"]},{"cell_type":"markdown","metadata":{},"source":["### 37. String & List - Reversing Transcription\n","#### Question: Write a function to reverse transcribe an mRNA sequence \"AUGCGUACGUU\" back into a DNA sequence."]},{"cell_type":"code","execution_count":33,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["MRNA : AUGCGUACGUU\n","DNA : ATGCGTACGTT\n"]}],"source":["rna = \"AUGCGUACGUU\"\n","dna = rna.replace('U','T')\n","\n","print(f'MRNA : {rna}')\n","print(f\"DNA : {dna}\")"]},{"cell_type":"markdown","metadata":{},"source":["### 38. List & String - Translating Overlapping Codons\n","#### Question: Write a function to translate overlapping codons in a DNA sequence \"ATGCGTACGTT\" by one base at a time."]},{"cell_type":"code","execution_count":65,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["{'A': 2, 'T': 4, 'G': 3, 'C': 2}\n"]}],"source":["def count_nuc(dna):\n","    count = {'A':0,'T':0,'G':0,'C':0}\n","\n","    for key in count.keys():\n","        count[key] = dna.count(key)\n","\n","    return count\n","\n","dna = \"ATGCGTACGTT\"\n","print(count_nuc(dna))"]},{"cell_type":"markdown","metadata":{},"source":["### 39. List & Conditional Statements - Detecting Frameshifts\n","#### Question: Write a function to detect frameshift mutations between two sequences \"ATGCGTACGTT\" and \"ATCGTACGTT\"."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 40. String & List - Extracting Introns and Exons\n","#### Question: Write a function to extract exons and introns from a DNA sequence \"ATGCGTACGTT\" with exons at positions [(0, 3), (4, 7)]."]},{"cell_type":"code","execution_count":64,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["DNA : ATGCGTACGTT\n","Exon Location :  [(0, 3), (4, 7)]\n","Introns Location :  [(8, 10)]\n","Exons  :  ATGC GTAC \n","Introns :  GTT "]}],"source":["#Assuming in (0,3) means  Exon locations are 0,1,2,3\n","\n","\n","dna = \"ATGCGTACGTT\"\n","n = len(dna)\n","exon_loc = [(0, 3), (4, 7)]\n","introns =[] \n"," #[  if exon_loc[i+1][0] != None]\n","\n","for i in range(len(exon_loc)):\n","    if i == (len(exon_loc)-1):\n","      introns.append((exon_loc[i][1]+1,n-1))\n","      break\n","    else :\n","       start = exon_loc[i][1]+1      \n","       end = (exon_loc[i+1][0])-1\n","       if end < start :\n","          continue \n","       \n","       #print(start,end)\n","       introns.append((start,end))   \n","\n","print(f\"DNA : {dna}\")\n","print(\"Exon Location : \",exon_loc) \n","print(\"Introns Location : \",introns)\n","\n","\n","exons_value = [dna[start:end+1] for (start,end) in exon_loc]\n","\n","introns_value = [dna[start:end+1] for (start,end) in introns]\n","\n","\n","\n","print(\"Exons  : \",end=\" \")\n","for intron in exons_value:\n","    print(f\"{intron}\",end=\" \")\n","\n","print()\n","print(\"Introns : \",end=\" \")\n","for intron in introns_value:\n","    print(f\"{intron}\",end=\" \")\n","\n"]},{"cell_type":"markdown","metadata":{},"source":["### 41. String & Dictionary - Codon Usage Frequency\n","\n","#### Question: Write a function that calculates the frequency of each codon in a DNA sequence \"ATGCGTACGTTATGCGT\" and returns a dictionary with the counts.\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Repeat"]},{"cell_type":"markdown","metadata":{},"source":["### 42. Nested Loops - Finding Longest ORF\n","\n","#### Question: Write a function to find the longest open reading frame (ORF) in the DNA sequence \"ATGCGTACGTTATGCGTTAA\" that starts with 'ATG' and ends with a stop codon.\n","\n"]},{"cell_type":"code","execution_count":72,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["DNA : ATGCGTACTGGTAAATGCGTACGTAGATGCGTACCTGAATGCGTACTGATGA\n","Longest ORF : ATGCGTACTGGTAA\n"]}],"source":["import re \n","\n","dna = dnaseq = \"ATGCGTACTGGTAAATGCGTACGTAGATGCGTACCTGAATGCGTACTGATGA\" \n","\n","stop_codon = ['TAA','TAG','TGA']\n","pattern = r\"ATG((?:[ATGC]{1})*?)(TAA|TAG|TGA)\"\n","\n","pattern = re.compile(pattern , re.IGNORECASE)\n","matches = pattern.finditer(seq)\n","\n","longest_ORF = \"\"\n","\n","for i,match in enumerate(matches,1):    \n","    #print(match.group(0))\n","    if len(match.group(1)) >= len(longest_ORF):\n","        longest_ORF = match.group(0)\n","        #print(f\"updated max ORF {i} : {match.group(1)}\")\n","\n","    #print(f\" ORF {i} : {match.group(1)}\")\n","\n","print(f\"DNA : {dna}\")\n","print(f\"Longest ORF : {longest_ORF}\")\n","    \n","    "]},{"cell_type":"markdown","metadata":{},"source":["### 43. List & Set - Unique Amino Acids in Protein Sequence\n","\n","#### Question: Write a function to identify unique amino acids in a protein sequence \"MKVLYRFY\" using a set.\n","\n","##### Output: {'R', 'V', 'K', 'L', 'Y', 'M', 'F'}"]},{"cell_type":"code","execution_count":74,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["{'L', 'Y', 'F', 'R', 'K', 'V', 'M'}\n"]}],"source":["protein = \"MKVLYRFY\"\n","uni_amino = set(protein)\n","\n","\n","print(uni_amino)"]},{"cell_type":"markdown","metadata":{},"source":["### 44. String Manipulation - Finding Overlapping K-mers\n","\n","#### Question: Write a function to find all overlapping k-mers of length 3 in the DNA sequence \"ATGCGTACGTT\".\n","\n"]},{"cell_type":"code","execution_count":78,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["DNA : ATGCGTACGTT\n","kmer (k = 3):  ATG TGC GCG CGT GTA TAC ACG CGT GTT "]}],"source":["\n","def kmer(seq,k):\n","    kmer_list = []\n","\n","    #dna = \"ATGCGTACGTT\"\n","\n","    for i in range(0,n):\n","        kmer = seq[i:i+3]\n","        if len(kmer)< k:\n","            break\n","\n","        kmer_list.append(kmer)\n","\n","    return kmer_list\n","\n","dna = \"ATGCGTACGTT\"\n","\n","print(f\"DNA : {dna}\")\n","print(\"kmer (k = 3): \",end=\" \")\n","\n","for kmer in kmer(dna,3):\n","    print(f\"{kmer}\",end=\" \")     \n"]},{"cell_type":"markdown","metadata":{},"source":["### 45. List & Conditional Statements - GC Content Windows\n","\n","#### Question: Write a function to calculate the GC content in non-overlapping windows of size 4 in the DNA sequence \"ATGCGTACGTT\".\n","\n","#####  Output: [50.0, 50.0, 25.0]\n"]},{"cell_type":"code","execution_count":96,"metadata":{},"outputs":[],"source":["# easy"]},{"cell_type":"markdown","metadata":{},"source":["### 46. List & String - Translating Codon Frame Shifts\n","\n","#### Question: Write a function to translate all possible frames (0, +1, +2) in the DNA sequence \"ATGCGTACGTT\".\n","\n","##### # Output: [['Methionine', 'Arginine', 'Tyrosine'], ['Unknown', 'Unknown', 'Unknown'], ['Unknown', 'Unknown', 'Unknown']]"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 47. String & Dictionary - Complementary RNA Strand\n","#### Question: Write a function to generate the complementary RNA strand for the sequence \"AUGCGUACGUU\".\n","##### Output: 'UACGCAUGCAA'"]},{"cell_type":"code","execution_count":83,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["RNA : AUGCGUACGUU\n","comp Rna : UACGCAUGCAA\n"]}],"source":["comp_rna_pair ={\"A\":\"U\",\"U\":\"A\",\"G\":\"C\",\"C\":\"G\"}\n","rna  = \"AUGCGUACGUU\"\n","comp_rna = \"\"\n","\n","for char in rna :\n","    comp_rna += comp_rna_pair[char]\n","\n","print(f\"RNA : {rna}\")\n","print(f\"comp Rna : {comp_rna}\")\n"]},{"cell_type":"markdown","metadata":{},"source":["### 49. List & Loops - Translating Reverse Complement\n","#### Question: Write a function to translate the reverse complement of a DNA sequence \"ATGCGTACGTT\" into a protein sequence.\n","\n","\n","##### # Output: ['Unknown', 'Unknown', 'Unknown', 'Unknown']"]},{"cell_type":"code","execution_count":95,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["ATGCGTACGTT\n","TACGCATGCAA\n","AACGTACGCAT\n","Asparagine-Valine-Arginine-Invalid-\n"]}],"source":["comp_pair_dna = {'A':'T','T':'A','G':'C','C':'G'}\n","dna = \"ATGCGTACGTT\"\n","\n","comp_dna = \"\"\n","for nuc in dna:\n","    comp_dna += comp_pair_dna[nuc]\n","rev_comp = comp_dna[::-1]\n","\n","print(dna)\n","print(comp_dna)\n","print(rev_comp)\n","\n","def dna_to_rna(dna):\n","    return dna.replace('T','U')\n","\n","\n","\n","codon_table = {\n","        'AUG': 'Methionine', 'UUU': 'Phenylalanine', 'UUC': 'Phenylalanine', 'UUA': 'Leucine', 'UUG': 'Leucine',\n","        'CUU': 'Leucine', 'CUC': 'Leucine', 'CUA': 'Leucine', 'CUG': 'Leucine', 'AUU': 'Isoleucine', 'AUC': 'Isoleucine',\n","        'AUA': 'Isoleucine', 'GUU': 'Valine', 'GUC': 'Valine', 'GUA': 'Valine', 'GUG': 'Valine', 'UCU': 'Serine',\n","        'UCC': 'Serine', 'UCA': 'Serine', 'UCG': 'Serine', 'CCU': 'Proline', 'CCC': 'Proline', 'CCA': 'Proline',\n","        'CCG': 'Proline', 'ACU': 'Threonine', 'ACC': 'Threonine', 'ACA': 'Threonine', 'ACG': 'Threonine', 'GCU': 'Alanine',\n","        'GCC': 'Alanine', 'GCA': 'Alanine', 'GCG': 'Alanine', 'UAU': 'Tyrosine', 'UAC': 'Tyrosine', 'UAA': 'Stop',\n","        'UAG': 'Stop', 'CAU': 'Histidine', 'CAC': 'Histidine', 'CAA': 'Glutamine', 'CAG': 'Glutamine', 'AAU': 'Asparagine',\n","        'AAC': 'Asparagine', 'AAA': 'Lysine', 'AAG': 'Lysine', 'GAU': 'Aspartic Acid', 'GAC': 'Aspartic Acid',\n","        'GAA': 'Glutamic Acid', 'GAG': 'Glutamic Acid', 'UGU': 'Cysteine', 'UGC': 'Cysteine', 'UGA': 'Stop',\n","        'UGG': 'Tryptophan', 'CGU': 'Arginine', 'CGC': 'Arginine', 'CGA': 'Arginine', 'CGG': 'Arginine', 'AGU': 'Serine',\n","        'AGC': 'Serine', 'AGA': 'Arginine', 'AGG': 'Arginine', 'GGU': 'Glycine', 'GGC': 'Glycine', 'GGA': 'Glycine',\n","        'GGG': 'Glycine'\n","    }\n","\n","def translate_rna_protein(codon : str):\n","    if codon in codon_table.keys():\n","        return codon_table[codon]\n","    else :\n","        return \"Invalid\"\n","    \n","def rna_to_protein(codon_seq):\n","    n = len(codon_seq)\n","\n","    protein_seq = \"\"\n","\n","    for i in range(0,n,3):     \n","        protein_seq = protein_seq + translate_rna_protein(codon_seq[i:i+3])+\"-\"\n","\n","    return protein_seq \n","\n","\n","    \n","\n","\n","result = rna_to_protein((dna_to_rna(rev_comp)))\n","\n","print(result)\n","\n","\n"]},{"cell_type":"markdown","metadata":{},"source":["### 50. List & String - Identifying Start and Stop Codons\n","\n","#### Question: Write a function to find all positions of start codons 'ATG' and stop codons ('TAA', 'TAG', 'TGA') in a DNA sequence \"ATGCGTATGCGTTAA\".\n","\n","##### # Output: ([0, 5], [11])"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["import re\n","\n"]},{"cell_type":"markdown","metadata":{},"source":["### 51. String & List - Generating K-mer Frequencies\n","\n","#### Question: Write a function to calculate the frequency of each k-mer of length 3 in the DNA sequence \"ATGCGTACGTT\".\n","\n","##### # Output: {'ATG': 1, 'TGC': 1, 'GCG': 1, 'CGT': 2, 'GTA': 1, 'TAC': 1, 'ACG': 1, 'GTT': 1}\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 52. Dictionary & List - Counting Dinucleotides\n","\n","#### Question: Write a function to count the frequency of each dinucleotide pair (e.g., 'AA', 'AC', etc.) in a DNA sequence \"ATGCGTACGTT\".\n","\n","##### # Output: {'AT': 1, 'TG': 1, 'GC': 2, 'CG': 2, 'GT': 2, 'TA': 1, 'AC': 1}\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 53. List & String - Finding Reverse Palindromes\n","\n","#### Question: Write a function to find all reverse palindromic sequences of length 6 in the DNA sequence \"ATGCGTACGCGTACGT\".\n","\n","##### Output: ['CGTACG', 'GCGTAC']"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 54. String & List - Protein Subsequence Search\n","\n","#### Question: Write a function to find all occurrences of a protein subsequence \"LYR\" in a protein sequence \"MKVLYRLYRFY\".\n","\n","##### Output: [3, 6]"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 55. List & Dictionary - Finding Codon Usage Bias\n","\n","#### Question: Write a function to compare codon usage in two different DNA sequences \"ATGCGTACGTT\" and \"ATGCGTAGCGT\".\n","\n","##### Output: ({'ATG': 1, 'CGT': 2, 'ACG': 1, 'GTT': 1}, {'ATG': 1, 'CGT': 2, 'AGC': 1, 'GTA': 1})\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 56. List & String - Converting DNA to Protein Using a Custom Genetic Code\n","\n","#### Question: Write a function to translate a DNA sequence \"ATGCGTACGTT\" into a protein using a custom genetic code mapping.\n","\n","##### Output: ['Methionine', 'Arginine', 'Tyrosine', 'Valine']\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["custom_genetic_code = {\n","    'ATG': 'Methionine', 'CGT': 'Arginine', 'TAC': 'Tyrosine', 'GTT': 'Valine',\n","    'TAA': 'Stop', 'TAG': 'Stop', 'TGA': 'Stop'\n","}"]},{"cell_type":"markdown","metadata":{},"source":["### 57. List & Set - Finding Unique Codons\n","\n","#### Question: Write a function to find all unique codons in a DNA sequence \"ATGCGTACGTTATGCGT\".\n","\n","##### Output: {'ATG', 'CGT', 'ACG', 'TTA'}"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 58. String & List - Translating Protein Sequences with Multiple Start Codons\n","\n","#### Question: Write a function to translate all possible protein sequences starting from each occurrence of 'ATG' in a DNA sequence \"ATGCGTATGCGTTAA\".\n","\n","##### Output: [['Methionine', 'Arginine', 'Tyrosine'], ['Methionine', 'Arginine', 'Valine']]\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 59. List & Dictionary - Calculating Codon Usage in Different Frames\n","\n","#### Question: Write a function to calculate codon usage in all three reading frames of a DNA sequence \"ATGCGTACGTT\".\n","\n","##### Output: [{'ATG': 1, 'CGT': 2, 'ACG': 1, 'GTT': 1}, {'TGC': 1, 'GTA': 1, 'CGT': 1, 'TAC': 1}, {'GCG': 1, 'TAC': 1, 'GT': 1}]\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 60. List & Conditional Statements - Identifying ORFs in Reverse Complement\n","\n","#### Question: Write a function to find all open reading frames (ORFs) in the reverse complement of a DNA sequence \"ATGCGTACGTT\" that start with 'ATG' and end with a stop codon.\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 61. String & Set - Finding Overlapping Motifs\n","\n","#### Question: Write a function to find all occurrences of the motif \"ACGT\" in the DNA sequence \"ATGCGTACGTTACGTACGT\".\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 62. List & Dictionary - Counting Triplet Nucleotide Repeats\n","\n","#### Question: Write a function to count the number of times each triplet nucleotide repeat occurs in the DNA sequence \"ATGCGTACGTTACG\".\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 63. String & Loops - Translating a Custom Genetic Code Sequence\n","\n","#### Question: Write a function to translate a DNA sequence \"ATGCGTACGTT\" using a custom genetic code that includes ambiguous codons (e.g., 'ATN' -> 'Methionine', 'CGN' -> 'Arginine').\n","\n"]},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[],"source":["genetic_code = {\n","    'ATG': 'Methionine', 'ATN': 'Methionine', 'CGT': 'Arginine', 'CGN': 'Arginine',\n","    'TAC': 'Tyrosine', 'GTT': 'Valine', 'TAA': 'Stop', 'TAG': 'Stop', 'TGA': 'Stop'\n","}"]},{"cell_type":"markdown","metadata":{},"source":["### 64. String & List - Analyzing Amino Acid Composition\n","\n","#### Question: Write a function to calculate the composition of each amino acid in a protein sequence \"MKVLYRFY\".\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 65. List & Dictionary - Calculating Nucleotide Composition in Codon Positions\n","\n","#### Question: Write a function to calculate the nucleotide composition at each codon position (1st, 2nd, 3rd) in the DNA sequence \"ATGCGTACGTT\".\n","\n","##### # Output: {1: {'A': 1, 'T': 1, 'C': 1, 'G': 2}, 2: {'A': 0, 'T': 1, 'C': 3, 'G': 1}, 3: {'A': 1, 'T': 2, 'C': 0, 'G': 2}}\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 66. List & Set - Identifying Non-overlapping Motifs\n","\n","#### Question: Write a function to find all non-overlapping occurrences of the motif \"ACG\" in the DNA sequence \"ATGCGTACGTTACG\".\n","\n","##### Output: [5, 11]"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 67. List & Dictionary - Creating a Reverse Complement Dictionary\n","\n","#### Question: Write a function to create a dictionary that maps each codon to its reverse complement in a DNA sequence \"ATGCGTACGTT\".\n","\n","##### Output: {'ATG': 'CAT', 'CGT': 'ACG', 'ACG': 'CGT', 'GTT': 'AAC'}\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 68. List & String - Transcribing DNA with Ambiguous Bases\n","\n","#### Question: Write a function to transcribe a DNA sequence with ambiguous bases \"ATGCGTNCGTT\" into an RNA sequence, where 'N' represents any nucleotide.\n","\n","##### Output: 'UACGCANGCAA'"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["### 69. String & List - Translating Overlapping Codons with Degenerate Bases\n","\n","#### Question: Write a function to translate overlapping codons in a DNA sequence \"ATGCGTACGTNNN\" using a genetic code that includes degenerate bases.\n","\n","##### Output: ['Methionine', 'Arginine', 'Threonine', 'Any']\n"]},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[],"source":["degenerate_genetic_code = {\n","    'ATG': 'Methionine', 'CGT': 'Arginine', 'ACG': 'Threonine', 'TNN': 'Any'\n","}"]},{"cell_type":"markdown","metadata":{},"source":["### 70. List & String - Finding Longest Protein Coding Sequence\n","\n","#### Question: Write a function to find the longest protein coding sequence in the DNA sequence \"ATGCGTACGTTGAAATGCCGTTAG\"."]},{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":["genetic_code = {\n","    'ATG': 'Methionine', 'CGT': 'Arginine', 'TAC': 'Tyrosine', 'GTT': 'Valine',\n","    'TAA': 'Stop', 'TAG': 'Stop', 'TGA': 'Stop'\n","}\n","\n"]},{"cell_type":"markdown","metadata":{},"source":["## Please use the below seq for verifying the results"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["dna_sequences = [\n","    \"ATGCGTACGTTGACGTAGCCTAGCGTACGATTACGCGTATGGGCTACTGCGTACGTTGCGTATGCGTACGTTGAATGCGT\",\n","    \"GCTAGCGTACGTTGCGTAGCGTACGTGACGTACTGCGTAGCTAGCGTTACGTTACGCGTACGATGCGTACGTGCGTGACG\",\n","    \"ATGCGTACGTTGCGTATGCGTACGTTGACGTAGCTAGCGTACGTTACGCGTACGCTGCGTAGCGTACGCGTATGCGTACG\",\n","    \"CGTACGTTGACGTAGCGTACGTGCGTACGCGTACGCTAGCGTACGTTGCGTACGTACGCGTACGTTGCGTACGCGTACGT\",\n","    \"ACGTTGCGTACGCGTACGTTGACGTAGCGTACGTTACGCGTACGTGCGTAGCGTACGCTGCGTACGCGTACGTTGCGTAC\"\n","]\n","\n","protein_sequences = [\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\",\n","    \"MVVVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\",\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\",\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\",\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","]"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["dna_sequences = [\n","    \"ATGCGTACGTTGACGTAGCCTAGCGTACGATTACGCGTATGGGCTACTGCGTACGTTGCGTATGCGTACGTTGAATGCGT\"\n","    \"TACGTTGACGTAGCGTACGTGCGTAGCTAGCGTACGTTACGCGTACGATGCGTACGTGCGTGACGTACGTTACGCGTACG\"\n","    \"TAGCGTACGTTGCGTACGTTGACGTAGCGTACGCTGCGTAGCGTACGCGTATGCGTACGTTGACGTAGCGTACGTGCGTA\"\n","    \"GCTAGCGTACGTTGCGTACGTACGCGTACGTTGCGTACGCGTACGTTGACGTAGCGTACGTTACGCGTACGTGCGTAGCG\",\n","    \n","    \"GCTAGCGTACGTTGCGTAGCGTACGTGACGTACTGCGTAGCTAGCGTTACGTTACGCGTACGATGCGTACGTGCGTGACG\"\n","    \"TACGTTGACGTAGCGTACGTTACGCGTACGATGCGTACGTGCGTGACGTACGTTACGCGTACGTGCGTAGCGTACGTTAC\"\n","    \"GCGTACGTGCGTACGTTGACGTAGCGTACGTTACGCGTACGTGCGTAGCGTACGCTGCGTACGCGTACGTTGCGTACGCG\"\n","    \"TACGTGCGTACGTTGACGTAGCGTACGTTACGCGTACGTGCGTAGCGTACGTTACGCGTACGTTGACGTAGCGTACGTTG\",\n","    \n","    \"ATGCGTACGTTGCGTATGCGTACGTTGACGTAGCTAGCGTACGTTACGCGTACGCTGCGTAGCGTACGCGTATGCGTACG\"\n","    \"GCGTACGTTGCGTAGCGTACGTTGACGTAGCGTACGCTGCGTAGCGTACGCGTATGCGTACGTTGACGTAGCGTACGTGC\"\n","    \"GTTACGCGTACGTGCGTAGCGTACGCTGCGTACGCGTACGTTGCGTACGTGCGTACGTTGACGTAGCGTACGTTACGCGT\"\n","    \"ACGTGCGTAGCGTACGTTGACGTAGCGTACGTTACGCGTACGTGCGTAGCGTACGTTGCGTACGCGTACGTTGACGTAGC\",\n","    \n","    \"CGTACGTTGACGTAGCGTACGTGCGTACGCGTACGCTAGCGTACGTTGCGTACGTACGCGTACGTTGCGTACGCGTACGT\"\n","    \"GCGTACGTTGACGTAGCGTACGTGCGTACGCGTACGTTGACGTAGCGTACGTTACGCGTACGCTGCGTACGCGTACGTTG\"\n","    \"CGTACGTTGACGTAGCGTACGTGCGTACGCGTACGCTAGCGTACGTTGCGTACGTACGCGTACGTTGCGTACGCGTACGT\"\n","    \"GCGTACGTTGACGTAGCGTACGTGCGTACGCGTACGTTGACGTAGCGTACGTTACGCGTACGCTGCGTACGCGTACGTTG\",\n","    \n","    \"ACGTTGCGTACGCGTACGTTGACGTAGCGTACGTTACGCGTACGTGCGTAGCGTACGCTGCGTACGCGTACGTTGCGTAC\"\n","    \"GCGTACGTGCGTACGTTGACGTAGCGTACGTTACGCGTACGTGCGTAGCGTACGCTGCGTACGCGTACGTTGCGTACGCG\"\n","    \"TACGTGCGTAGCGTACGTTGACGTAGCGTACGTTACGCGTACGTGCGTAGCGTACGCTGCGTACGCGTACGTTGCGTACG\"\n","    \"GTTGACGTAGCGTACGTTACGCGTACGTGCGTAGCGTACGCTGCGTACGCGTACGTTGCGTACGCGTACGTGCGTAGCGT\"\n","]\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["protein_sequences = [\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\",\n","    \n","    \"MVVVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\",\n","    \n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\",\n","    \n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\",\n","    \n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","    \"MKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFYMKVLYRLYRFY\"\n","]\n"]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.12.4"}},"nbformat":4,"nbformat_minor":0}
