{"cells":[{"cell_type":"markdown","metadata":{"id":"rpXCdX30NNRx"},"source":["# Q1\n","### Write a Python program that classifies a given protein sequence into one of five enzyme classes based on specific hypothetical motifs. Each enzyme class is associated with a unique motif: Hydrolase (GXSXG), Transferase (KTXN), Oxidoreductase (CXXC), Lyase (DXXD), and Ligase (RXYK). \n","### The program should check if any of these motifs are present in the input sequence and print the corresponding enzyme class. If no motifs are found, the program should indicate that no matching enzyme class exists"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"bxlAtYMAjZQS"},"outputs":[],"source":["#Write your code here"]},{"cell_type":"markdown","metadata":{"id":"ksBaOdlAWIw4"},"source":["# Q2\n","### Write a Python program that checks if a given amino acid sequence is valid.The program should determine if the sequence contains any invalid amino acids, defined as 'B', 'J', 'O', 'U', 'X', and 'Z'.\n","### If an invalid amino acid is found, the program should print the first invalid amino acid and state that the sequence is invalid. If the sequence contains only valid amino acids (A, C, D, E, F, G, H, I, K, L, M, N, P, Q, R, S, T, V, W, Y), the program should confirm that the sequence is valid."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"kFeg2eKgjyjx"},"outputs":[],"source":["#Write your code here"]},{"cell_type":"markdown","metadata":{"id":"TQ8q_WsO1aDH"},"source":["# Q3\n","### To determine the ATGC content of a given DNA sequence, count the occurrences of each nucleotide and calculate their percentage content relative to the total sequence length.\n","\n","### For the sequence \"ATGCGATAACTAG\", the contents are approximately 30.77% Adenine, 23.08% Thymine, 15.38% Guanine, and 23.08% Cytosine."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"UAYTefK7kBBW"},"outputs":[],"source":["#Write your code here"]},{"cell_type":"markdown","metadata":{"id":"xZEj7V2Z2VZA"},"source":["# Q4\n","How can we construct the reverse complement of a given DNA sequence?\n","\n","### Explanation\n","The reverse complement of a DNA sequence is obtained by two steps:\n","\n","### Complement: Replace each nucleotide in the sequence with its complement:\n","* Adenine (A) is replaced with Thymine (T)\n","* Thymine (T) is replaced with Adenine (A)\n","* Guanine (G) is replaced with Cytosine (C)\n","* Cytosine (C) is replaced with Guanine (G)\n","\n","### Reverse: Reverse the entire sequence after finding its complement.\n","Example:\n","Consider the following DNA sequence: \"ATGCGATAACTAG\"\n","\n","Complement:\n","\n","* A -> T\n","* T -> A\n","* G -> C\n","* C -> G\n","* G -> C\n","* A -> T\n","* T -> A\n","* A -> T\n","* A -> T\n","* C -> G\n","* T -> A\n","* A -> T\n","* G -> C\n","\n","* Complemented sequence: \"TACGCTATTGATC\"\n","* Reverse:\n","Reversed complemented sequence: \"CTAGTTATCGCAT\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"DB72-PVckJqt"},"outputs":[],"source":["#Write your code here"]},{"cell_type":"markdown","metadata":{"id":"RiWQZZvz3iIx"},"source":["# Q5.\n","How can we calculate the melting temperature (Tm) of a given DNA sequence?\n","\n","### Explanation\n","The melting temperature (Tm) of a DNA sequence is the temperature at which half of the DNA duplex dissociates to become single-stranded. It is a crucial parameter in various molecular biology techniques such as PCR.\n","\n","### There are different methods to estimate Tm, but a commonly used method for short oligonucleotides is the Wallace rule:\n","\n","### For sequences shorter than 14 nucleotides:\n","ùëáùëö=2√ó(number¬†of¬†A/T)+4√ó(number¬†of¬†G/C)\n","* For longer sequences, more complex formulas involving the sequence length and the concentration of ions (salt) in the solution are used, but the Wallace rule provides a quick and useful approximation.\n","\n","Count the Nucleotides:\n","\n","Adenine (A): 4\n","Thymine (T): 3\n","Guanine (G): 2\n","Cytosine (C): 3\n","Apply the Wallace Rule:\n","\n","Number of A/T: 7\n","Number of G/C: 5\n","ùëáùëö=2√ó7+4√ó5=14+20=34‚àòùê∂\n","Tm=2√ó7+4√ó5=14+20=34‚àòC\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ZTaS8JgbkTwU"},"outputs":[],"source":["#Write your code here"]},{"cell_type":"markdown","metadata":{"id":"ZW-IAG4SHtEA"},"source":["# Q6\n","### For a given sequence, find the most occurring K-mer count.\n","### Part 1: Finding the Most Occurring k-mer with k=3\n","#### Given the nucleotide sequence: ATGCGTACGTAGCTAGCTAGCTGACGTAGCTAGCTACGATCGTACGTACGATCGTACGTAGCTAGCTACGTAGCTAGCTAGCTACGTAGCTAGCTAGCTAGCTG\n","#### Use the kmer_count function to find the most occurring k-mer of size 3. What is the k-mer and its count in this sequence?\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"zWAXuZsSkpR_"},"outputs":[],"source":["#write your code here"]},{"cell_type":"markdown","metadata":{"id":"WWaX-YQ5ktfX"},"source":["Part 2: Finding the Most Occurring k-mer with k Between 10 and 15\n","Given the same nucleotide sequence, Using the kmer_count function, determine the most occurring k-mer for each k value between 10 and 15. List the k-mer and its count for each value of k."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"HWAdg0GBkx3c"},"outputs":[],"source":["#Write your code here"]},{"cell_type":"markdown","metadata":{"id":"Dod77IjNzM2T"},"source":["## Q7.\n","Given a chromosome sequence, identify the number of genes present in it by locating the start and stop codons within the sequence. How can this be achieved, and what is the count of genes in the provided sequence?\n","Hint:\n","To determine the number of genes in a chromosome sequence, we need to locate the start and stop codons that define the boundaries of each gene. In the context of DNA sequences, the start codon is typically \"ATG\" and the stop codons are \"TAA\", \"TAG\", and \"TGA\". The process involves scanning the sequence for these codons in all three reading frames and counting the number of valid genes.\n","\n","### Here's a step-by-step approach:\n","\n","1. Identify Start Codons: Search for the \"ATG\" codon in the sequence.\n","2. Identify Stop Codons: Search for the \"TAA\", \"TAG\", and \"TGA\" codons in the sequence.\n","3. Reading Frames: Since DNA can be read in three frames, repeat the search in all three frames.\n","4. Gene Validation: For each start codon, find the nearest downstream stop codon in the same frame. If found, count it as a valid gene.\n","\n","Example\n","Consider the following chromosome sequence: \"ATGCGATAACTAGATGTGA\"\n","\n","* Frame 1:\n","Start at position 0: \"ATG\"\n","Stop at position 6: \"TAA\", position 15: \"TGA\"\n","Gene: \"ATGCGATAA\" (from position 0 to 8)\n","* Frame 2:\n","No start codons at positions 1, 4, or 7.\n","* Frame 3:\n","Start at position 4: \"ATG\"\n","Stop at position 10: \"TAG\", position 15: \"TGA\"\n","Gene: \"ATGTGA\" (from position 4 to 15)\n","* Total number of genes: 2\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"V25FGu6SlB05"},"outputs":[],"source":["#Write your code here"]},{"cell_type":"markdown","metadata":{},"source":["# Q8\n","### Write a Python program that identifies all non-overlapping occurrences of a DNA motif in a given sequence. The motif consists of substrings composed of 'A' and/or 'T' with lengths between 3 and 6 characters. \n","### The program should return each motif found along with its starting and ending indices in the sequence.\n","#### Input DNA sequence = 'AATGAAGGGCCGCTACGATAAGGAACTTCGTAATTTCAG'"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["#write your code here"]},{"cell_type":"markdown","metadata":{},"source":["# Q9\n","### Open reading frames (ORFs) are regions of DNA that can be translated into proteins. An ORF starts with a start codon (such as ATG) and ends with one of the stop codons (such as TAA, TAG, or TGA). Write a Python function using the re library to identify all ORFs in a given DNA sequence. Your function should return the starting and ending indices (0-based) of each ORF.\n","\n","## Details:\n","\n","### The DNA sequence is a string consisting of characters 'A', 'C', 'G', and 'T'. The start codon is always ATG. The stop codons are TAA, TAG, and TGA.\n","\n","#### Input seq = 'ATGCGATCGACGCTAGCGATCGCGATCGATGGCGATCGCTAGCGATCGATCGCGATCGTAAAGGCTACGTGTCAGTAA'"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["#write your code here"]},{"cell_type":"markdown","metadata":{},"source":["# Q10\n","### In bioinformatics, the analysis of DNA sequences is fundamental to understanding genetic similarities and variations among different organisms or samples. One common task is to identify regions within DNA sequences that are conserved across different species or individuals. These conserved regions, known as motifs or subsequences, can provide insights into functional or evolutionary relationships. One method to find such regions is by identifying the Longest Common Subsequence (LCS) shared by multiple DNA sequences.\n","\n","### Problem:\n","#### Given a list of DNA sequences, you are required to write a Python function to find the longest common subsequence (LCS) shared by all the sequences in the list. The LCS is the longest sequence of nucleotides that appears in the same relative order (but not necessarily contiguously) in all the given DNA sequences.\n","\n","### Example, Suppose you have the following DNA sequences:\n","\n","* \"ATGCTGAC\"\n","* \"CTGACGT\"\n","* \"TGAC\"\n","* The LCS shared by all these sequences is \"TGAC\".\n","* Task:\n","### Q. Write a Python function lcs(sequences) that takes a list of DNA sequences as input and returns the longest common subsequence shared by all sequences in the list.\n","* Input sequences - \"ATGCGTACCGTACGTAGCTAGCTGACTGACGTAGCGTACGCTAGCTG\"\n","* \"CGTACGTAGCTAGCTGACTGACGTAGCGTACGCTAGCTGACGCTGAC\"\n","* \"GCTAGCTGACTGACGTAGCGTACGCTAGCTGACGCTGACTGCGTAGC\"\n","* \"TACGTAGCTAGCTGACTGACGTAGCGTACGCTAGCTGACGCTGACTG\"\n","* \"GTACGTAGCTAGCTGACTGACGTAGCGTACGCTAGCTGACGCTGACT\""]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["# Q11\n","#### Identifying Significant Changes in Protein Abundance\n","### In proteomics, researchers often measure protein abundance under different experimental conditions to understand how treatments, mutations, or environmental factors affect protein levels. To analyze these changes, one can compare the abundance of proteins between conditions and identify those with significant increases or decreases. For each protein, it's crucial to determine not just if its abundance has changed, but whether it has increased or decreased.\n","### You are provided with a list of tuples, where each tuple represents the abundance levels of a protein in two different conditions. Each tuple contains the protein identifier and its abundance levels in Condition 1 and Condition 2. You need to write a Python function that analyzes these tuples to determine which proteins are significantly different based on a given fold change threshold. Additionally, specify whether each protein is \"increased\" or \"decreased\" in Condition 2 compared to Condition 1.\n","Input:\n","### A list of tuples where each tuple is of the form (protein_id, abundance1, abundance2), where:\n","* protein_id is a unique identifier for the protein.\n","* abundance1 is the abundance level of the protein in Condition 1.\n","* abundance2 is the abundance level of the protein in Condition 2.\n","# A fold change threshold.\n","* Output:\n","### A list of tuples where each tuple is of the form (protein_id, change_type), where:\n","* protein_id is the unique identifier for the protein.\n","* change_type is either 'increased' or 'decreased', depending on whether the protein's abundance in Condition * 2 is higher or lower compared to Condition 1, respectively.\n","\n","#### Example:\n","\n","### Given the following list of tuples and a fold change threshold of 1.5:\n","\n","protein_abundance = [\n","    ('ProteinX', 5, 8),\n","    ('ProteinY', 20, 10),\n","    ('ProteinZ', 150, 300),\n","    ('ProteinW', 30, 15),\n","    ('ProteinV', 100, 100)\n","]\n","Task:\n","### Write a Python function analyze_protein_changes(protein_abundance, threshold) that takes the list of tuples and a fold change threshold as input and returns the list of tuples indicating the protein identifier and its abundance change status ('increased', 'decreased', or 'no change')."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"markdown","metadata":{},"source":["# Q12\n","### In genetics, the codon table is used to translate RNA sequences into proteins. Each codon, a sequence of three RNA nucleotides, corresponds to a specific amino acid or a stop signal during protein synthesis.\n","\n","### Write a Python function called translate_rna_to_protein that takes a string representing an RNA sequence as input and returns the corresponding protein sequence. Use a dictionary to map RNA codons to their corresponding amino acids. For simplicity, assume the RNA sequence is always divisible by 3 and contains only the characters 'A', 'U', 'C', and 'G'.\n","\n","``` #The codon table is given below:\n","AUG: 'Methionine', UUU: 'Phenylalanine', UUC: 'Phenylalanine', UUA: 'Leucine', UUG: 'Leucine',\n","CUU: 'Leucine', CUC: 'Leucine', CUA: 'Leucine', CUG: 'Leucine', AUU: 'Isoleucine', AUC: 'Isoleucine',\n","AUA: 'Isoleucine', GUU: 'Valine', GUC: 'Valine', GUA: 'Valine', GUG: 'Valine', UCU: 'Serine',\n","UCC: 'Serine', UCA: 'Serine', UCG: 'Serine', CCU: 'Proline', CCC: 'Proline', CCA: 'Proline',\n","CCG: 'Proline', ACU: 'Threonine', ACC: 'Threonine', ACA: 'Threonine', ACG: 'Threonine', GCU: 'Alanine',\n","GCC: 'Alanine', GCA: 'Alanine', GCG: 'Alanine', UAU: 'Tyrosine', UAC: 'Tyrosine', UAA: 'Stop',\n","UAG: 'Stop', CAU: 'Histidine', CAC: 'Histidine', CAA: 'Glutamine', CAG: 'Glutamine', AAU: 'Asparagine',\n","AAC: 'Asparagine', AAA: 'Lysine', AAG: 'Lysine', GAU: 'Aspartic Acid', GAC: 'Aspartic Acid',\n","GAA: 'Glutamic Acid', GAG: 'Glutamic Acid', UGU: 'Cysteine', UGC: 'Cysteine', UGA: 'Stop',\n","UGG: 'Tryptophan', CGU: 'Arginine', CGC: 'Arginine', CGA: 'Arginine', CGG: 'Arginine', AGU: 'Serine',\n","AGC: 'Serine', AGA: 'Arginine', AGG: 'Arginine', GGU: 'Glycine', GGC: 'Glycine', GGA: 'Glycine',\n","GGG: 'Glycine' ```\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["codon_table = {\n","    'TCA': 'S',    # Serina\n","    'TCC': 'S',    # Serina\n","    'TCG': 'S',    # Serina\n","    'TCT': 'S',    # Serina\n","    'TTC': 'F',    # Fenilalanina\n","    'TTT': 'F',    # Fenilalanina\n","    'TTA': 'L',    # Leucina\n","    'TTG': 'L',    # Leucina\n","    'TAC': 'Y',    # Tirosina\n","    'TAT': 'Y',    # Tirosina\n","    'TAA': '*',    # Stop\n","    'TAG': '*',    # Stop\n","    'TGC': 'C',    # Cisteina\n","    'TGT': 'C',    # Cisteina\n","    'TGA': '*',    # Stop\n","    'TGG': 'W',    # Triptofano\n","    'CTA': 'L',    # Leucina\n","    'CTC': 'L',    # Leucina\n","    'CTG': 'L',    # Leucina\n","    'CTT': 'L',    # Leucina\n","    'CCA': 'P',    # Prolina\n","    'CCC': 'P',    # Prolina\n","    'CCG': 'P',    # Prolina\n","    'CCT': 'P',    # Prolina\n","    'CAC': 'H',    # Histidina\n","    'CAT': 'H',    # Histidina\n","    'CAA': 'Q',    # Glutamina\n","    'CAG': 'Q',    # Glutamina\n","    'CGA': 'R',    # Arginina\n","    'CGC': 'R',    # Arginina\n","    'CGG': 'R',    # Arginina\n","    'CGT': 'R',    # Arginina\n","    'ATA': 'I',    # Isoleucina\n","    'ATC': 'I',    # Isoleucina\n","    'ATT': 'I',    # Isoleucina\n","    'ATG': 'M',    # Methionina\n","    'ACA': 'T',    # Treonina\n","    'ACC': 'T',    # Treonina\n","    'ACG': 'T',    # Treonina\n","    'ACT': 'T',    # Treonina\n","    'AAC': 'N',    # Asparagina\n","    'AAT': 'N',    # Asparagina\n","    'AAA': 'K',    # Lisina\n","    'AAG': 'K',    # Lisina\n","    'AGC': 'S',    # Serina\n","    'AGT': 'S',    # Serina\n","    'AGA': 'R',    # Arginina\n","    'AGG': 'R',    # Arginina\n","    'GTA': 'V',    # Valina\n","    'GTC': 'V',    # Valina\n","    'GTG': 'V',    # Valina\n","    'GTT': 'V',    # Valina\n","    'GCA': 'A',    # Alanina\n","    'GCC': 'A',    # Alanina\n","    'GCG': 'A',    # Alanina\n","    'GCT': 'A',    # Alanina\n","    'GAC': 'D',    # Acido Aspartico\n","    'GAT': 'D',    # Acido Aspartico\n","    'GAA': 'E',    # Acido Glutamico\n","    'GAG': 'E',    # Acido Glutamico\n","    'GGA': 'G',    # Glicina\n","    'GGC': 'G',    # Glicina\n","    'GGG': 'G',    # Glicina\n","    'GGT': 'G'     # Glicina\n","}"]},{"cell_type":"markdown","metadata":{},"source":["# Q13\n","### In genetics, mutations refer to changes in the DNA sequence. Given a reference DNA sequence and a mutated DNA sequence, write a Python function called find_mutations that identifies the positions where mutations have occurred. The function should take two strings as input: the reference DNA sequence and the mutated DNA sequence (both of the same length), and return a dictionary where the keys are the positions (0-based index) of the mutations and the values are tuples containing the reference nucleotide and the mutated nucleotide.\n","\n","## Input sequence:\n","#### reference_sequence = \"ATCGGCTAATCGGCTAGCTAGCTAGCTGATCGGCTAATCGGCTAGCTAGCTAGCTGATCGGCTAATCGGCTAGCTAGCTAGCTGATCGGCTAATCGGCTAGCTAGCTAGCTG\"\n","\n","#### mutated_sequence = \"ATCGGCTGATCGGCTTGCTAGCTGGCTGATCGGCTAATCGGCTAGCTAGCTGGCTGATCGGCTAATCGGCTAGCTTGCTAGCTGATCGGCTAATCGGCTAGCTAGCTAGCAG\""]},{"cell_type":"markdown","metadata":{},"source":[]}],"metadata":{"colab":{"authorship_tag":"ABX9TyNVN9FKnlhbK0fhLfVXg7T5","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}
